src/tool.rs

```diff
@@ -74,70 +91,102 @@ impl Tool {
         cuda: bool,
         cached_compiler_family: &Mutex<HashMap<Box<Path>, ToolFamily>>,
         cargo_output: &CargoOutput,
+        out_dir: Option<&Path>,
     ) -> Self {
-        fn detect_family_inner(path: &Path, cargo_output: &CargoOutput) -> ToolFamily {
-            let mut cmd = Command::new(path);
-            cmd.arg("--version");
-
-            let stdout = match run_output(
-                &mut cmd,
+        fn is_zig_cc(path: &Path, cargo_output: &CargoOutput) -> bool {
+            run_output(
+                Command::new(&path).arg("--version"),
                 path,
                 // tool detection issues should always be shown as warnings
                 cargo_output,
             )
-            .ok()
-            .and_then(|o| String::from_utf8(o).ok())
-            {
-                Some(s) => s,
-                None => {
-                    // --version failed. fallback to gnu
-                    cargo_output.print_warning(&format_args!("Failed to run: {:?}", cmd));
-                    return ToolFamily::Gnu;
+            .map(|o| String::from_utf8_lossy(&o).contains("ziglang"))
+            .unwrap_or_default()
+        }
+
+        fn detect_family_inner(
+            path: &Path,
+            cargo_output: &CargoOutput,
+            out_dir: Option<&Path>,
+        ) -> Result<ToolFamily, Error> {
+            let tmp = NamedTempfile::new(
+                &out_dir
+                    .map(Cow::Borrowed)
+                    .unwrap_or_else(|| Cow::Owned(env::temp_dir())),
+                "detect_compiler_family.c",
+            )?;
+            tmp.file()
+                .write_all(include_bytes!("detect_compiler_family.c"))?;
```

Comment:
Is copying this to a temporary file necessary? Could you use `env!("CARGO_MANIFEST_DIR")/src/detect_compiler_family.c")` instead?

Reply:
That will only work if cc is used in `build.rs`

cc is also used in binary crates which has no access to the source code, so I'm afraid it won't work.

Reply:
`env!` is evaluated when the file containing the macro is compiled, not when the code is executed. So I think it could work.

Reply:
Yes, but what if the source of cc is not available at all?

What if cc-rs is used in binary crate, and re-distribute without including cc-rs source files?

Reply:
Why would being used in a library versus binary crate make a difference?

Reply:
Imagine this:

A binary crate pulls in cc-rs, compiles it into a single binary, and then distribute it to the user, who runs in on their computer.

`env!("CARGO_MANIFEST_DIR")` gives you the manifest dir for the source of cc-rs at compile-time, but it is not available in runtime, on the user's computer.

So we must use `include_bytes!` to copy the file into the final binary, and then write it to a named tempfile.

Reply:
Yeah, the approach here is regrettably necessary.